### PR TITLE
qa/cephadm: Add local dockerhub mirror

### PR DESCRIPTION
**Pull request by sebastian-philipp**
turned out to be more hacky than anticipated. 

Note that I only did run pytest, not the actual code.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: qa/tasks/cephadm.py [hunk 1]
```diff
@@ -1133,3 +1134,49 @@ def task(ctx, config):
...
+                } for reg in search
+            ]
+        }
+    else:
+        v2 = config
+    dockers = [r for r in v2['registry'] if r['prefix'] == 'docker.io']
+    if dockers:
+        docker = dockers[0]
+        docker['mirror'] = [{
+            "location": "vossi04.front.sepia.ceph.com:5000",
```
Comment (tchaikov) [inline qa/tasks/cephadm.py:1163]: might want to `s/vossi04.front.sepia.ceph.com:5000/mirror/` here?

File: qa/tasks/cephadm.py [hunk 2]
```diff
@@ -1133,3 +1134,49 @@ def task(ctx, config):
...
+
+def add_mirror_to_cluster(ctx, mirror):
+    # add other hosts
+    for remote in ctx.cluster.remotes.keys():
+        log.info('Add caching registry')
+        config = teuthology.get_file(
+            remote=remote,
+            path='/etc/containers/registry.conf'
+        )
+        new_config = toml.dumps(add_mirror(config))
```
Comment (tchaikov) [inline qa/tasks/cephadm.py:1176]: might want to pass `mirror` to `add_mirror()`.

File: qa/tasks/cephadm.py [hunk 3]
```diff
@@ -1133,3 +1134,49 @@ def task(ctx, config):
 
         finally:
             log.info('Teardown begin')
+
+
+def add_mirror(s, mirror):
+    import toml
```
Comment (kshtsk) [inline qa/tasks/cephadm.py]: This line should go to the top of the file

File: qa/tasks/cephadm.py [hunk 4]
```diff
@@ -11,6 +11,8 @@
 import uuid
 
 from io import BytesIO
+from typing import MutableMapping, Any
+import toml
```
Comment (tchaikov) [inline qa/tasks/cephadm.py:15]: as discussed we also need to install `toml` on sepia. might want to get the PRs for teuthology targeting master and py2 merged first.

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:15]: https://github.com/ceph/teuthology/pull/1493

File: qa/tasks/cephadm.py [hunk 5]
```diff
@@ -305,6 +305,7 @@ def ceph_bootstrap(ctx, config):
     ctx.cluster.run(args=[
         'sudo', 'chmod', '777', '/etc/ceph',
         ]);
+    add_mirror_to_cluster(ctx, config.get('docker_registry_mirror', 'vossi04.front.sepia.ceph.com:5000'))
```
Comment (kshtsk) [inline qa/tasks/cephadm.py:310]: can the vossi04* be declared as a constant at the beginning of the file

**Comment by tchaikov** [comment]
once the related commits are squashed and the teuthology is readied. this PR is good to go.

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:310]: this constant belongs to sepia, not into ceph/qa. but I don't know how.

File: qa/tasks/cephadm.py (hunk 2 above)
Comment (kshtsk) [inline qa/tasks/cephadm.py]: This overall code should go to teuthology repo as something like `teuthology.util.podman`

File: qa/tasks/cephadm.py (hunk 4 above)
Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:15]: is there a quick way to verify the code, before merging any PRs?

Comment (tchaikov) [reply qa/tasks/cephadm.py:15]: thanks, we will need to backport this change to octopus once the mirror is up and running.

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py:310]: I mean that should not be like a hard coded parameter.

File: qa/tasks/cephadm.py (hunk 4 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py:15]: > is there a quick way to verify the code, before merging any PRs?

you can run teuthology-suite with --teuthology-branch command to test it

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (tchaikov) [reply qa/tasks/cephadm.py:310]: i see. it's like a magic number.

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:309]: and it's actually **wrong** for other teuthology deployments.

File: qa/tasks/cephadm.py [hunk 6]
```diff
@@ -1133,3 +1135,51 @@ def task(ctx, config):
...
+        docker = dockers[0]
+        docker['mirror'] = [{
+            "location": mirror,
+            "insecure": True,
+        }]
+    return v2
+
+
+def add_mirror_to_cluster(ctx, mirror: str):
+    # add other hosts
```
Comment (kshtsk) [inline qa/tasks/cephadm.py:1171]: Add parameter descriptions

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:1171]: why?

File: qa/tasks/cephadm.py [hunk 7]
```diff
@@ -1133,3 +1135,51 @@ def task(ctx, config):
 
         finally:
             log.info('Teardown begin')
+
+
+def add_mirror(s, mirror: str):
```
Comment (kshtsk) [inline qa/tasks/cephadm.py:1140]: Maybe the proper name should be like `add_containers_registry`?

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:1140]: what about `registries_add_mirror_to_docker_io` ?

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py:309]: The customized values should be controlled by yaml files in a suite. You need to raise an exception or just ignore adding registry if corresponding key is missing.

Comment (kshtsk) [reply qa/tasks/cephadm.py:309]: also, could you rename it to `docker_registry` or `extra_docker_registry` if it does not override the default registry list in the `/etc/containers`.

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:309]: hm. this modifies the existing  "docker.io" registry by adding a new mirror. This is **not** another extra registry.

Comment (sebastian-philipp) [reply qa/tasks/cephadm.py:309]: > The customized values should be controlled by yaml files in a suite. You need to raise an exception or just ignore adding registry if corresponding key is missing.

no clue where this yaml file is located.

File: qa/tasks/cephadm.py [hunk 8]
```diff
@@ -1133,3 +1135,51 @@ def task(ctx, config):
...
+    if dockers:
+        docker = dockers[0]
+        docker['mirror'] = [{
+            "location": mirror,
+            "insecure": True,
+        }]
+    return v2
+
+
+def add_mirror_to_cluster(ctx, mirror: str):
```
Comment (kshtsk) [inline qa/tasks/cephadm.py:1170]: if you named it as `add_mirror_to_cluster` then signature should be `(mirror: str, cluster: teuthology.orchestra.cluster)`?
But it is better to give the name `add_containers_registry_mirror(mirror: str, cluster)` or something.

File: qa/tasks/cephadm.py (hunk 7 above)
Comment (kshtsk) [inline qa/tasks/cephadm.py:1140]: Also, add doc to the method with parameter description

Comment (kshtsk) [reply qa/tasks/cephadm.py]: Hm.. the code looks like do not add any mirror it just generations a another string based on `s`. And it actually not adding (appending, prepending) it is just set, so maybe:
`_override_containers_registry_mirror` or so.

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py:309]: > hm. this modifies the existing "docker.io" registry by adding a new mirror. This is **not** another extra registry.

so maybe don't use `add` keyword in method names it does misguide

File: qa/tasks/cephadm.py [hunk 9]
```diff
@@ -1133,3 +1135,52 @@ def task(ctx, config):
...
             log.info('Teardown begin')
+
+
+def registries_add_mirror_to_docker_io(conf, mirror):
+    config = toml.loads(conf)
+    is_v1 = 'registries' in config
+    if is_v1:
+        search = config.get('registries', {}).get('search', {}).get('registries', [])
+        insecure = config.get('registries', {}).get('search', {}).get('insecure', [])
+        # v2: MutableMapping[str, Any] = { needs Python 3
```
Comment (kshtsk) [inline qa/tasks/cephadm.py:1146]: why this is required?

File: qa/tox.ini [hunk 10]
```diff
@@ -23,3 +23,12 @@ commands = mypy {posargs:.}
 basepython = python3
 deps = {env:TEUTHOLOGY_GIT:git+https://github.com/ceph/teuthology.git@master}#egg=teuthology[coverage,orchestra,test]
 commands = python test_import.py {posargs:tasks/**/*.py}
+
+[testenv:pytest]
+basepython = python3
+deps =
+  {env:TEUTHOLOGY_GIT:git+https://github.com/ceph/teuthology.git@master}#egg=teuthology[test]
+  pytest
+  toml
```
Comment (kshtsk) [inline qa/tox.ini:32]: this will not be required as soon as teuthology has this change

File: qa/tox.ini [hunk 11]
```diff
@@ -23,3 +23,12 @@ commands = mypy {posargs:.}
 basepython = python3
 deps = {env:TEUTHOLOGY_GIT:git+https://github.com/ceph/teuthology.git@master}#egg=teuthology[coverage,orchestra,test]
 commands = python test_import.py {posargs:tasks/**/*.py}
+
+[testenv:pytest]
+basepython = python3
+deps =
+  {env:TEUTHOLOGY_GIT:git+https://github.com/ceph/teuthology.git@master}#egg=teuthology[test]
+  pytest
```
Comment (kshtsk) [inline qa/tox.ini:31]: I thought pytest is already a dependency for teuthology[test].

File: qa/tasks/cephadm.py (hunk 5 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py:309]: it can be in any *suitable* under the directory `qa/suites/rados/cephadm/`

File: qa/tasks/cephadm.py (hunk 7 above)
Comment (kshtsk) [reply qa/tasks/cephadm.py]: I don't see any `docker.io` word in the package which provides this config file, so maybe it is not good suffix. Package is `libcontainers`, the file path is `/etc/containers/registries.conf`, and as we figured out in other comment it is not _added_ it is more like _set_. So maybe `_set_containers_registry_mirror` and if mirrors accept a list, why not use plural and support lists?

**Comment by sebastian-philipp** [comment]
http://pulpito.front.sepia.ceph.com/swagner-2020-05-26_13:36:40-rados:cephadm-wip-swagner-testing-2020-05-25-0943-distro-basic-smithi/

succeeded. more or less. but in any case, we're blocked without this PR

**Pull request merged**